Introduce Y022/Y023/Y024/Y025: Imports linting error codes

AlexWaygood opened the pull request:
Helps work towards #80 and #32. I've tried to keep these error codes completely cross-compatible between Python 2 and Python 3 -- I think it would be best to leave Python-3-only changes (e.g. disallowing `typing.ContextManager` and `typing_extensions.OrderedDict`) can be left for another error code.

JelleZijlstra commented on README.rst:
```diff
@@ -81,6 +81,14 @@ currently emitted:
   an instance of ``cls``, and ``__new__`` methods.
 * Y020: Quoted annotations should never be used in stubs.
 * Y021: Docstrings should not be included in stubs.
+* Y022: Imports linting: use typing-module aliases to stdlib objects as little
+  as possible (e.g. use ``builtins.list`` instead of ``typing.List``). However,
+  use ``typing.NamedTuple`` instead of ``collections.namedtuple``, as it allows
+  for more precise type inference. For typing-only objects such as ``ClassVar``
+  or ``NoReturn``, import objects from ``typing`` instead of
+  ``typing_extensions`` wherever possible.
```
I feel like this should be split in three:
- one code for PEP 585 stuff (builtins instead of typing)
- one for NamedTuple
- one for using typing instead of typing-extensions where applicable

AlexWaygood replied:
It's annoying because from the perspective of writing the code, it's the same check. But I see what you mean, I'll revise it 🙂

AlexWaygood commented:
Okay, should be ready for review now. I'll hold off running a codemodding script on the Python-2 parts of typeshed until people have had a chance to review :)

JelleZijlstra commented on pyi.py:
```diff
@@ -42,6 +42,58 @@ class TypeVarInfo(NamedTuple):
+#
+# ChainMap does not exist in typing or typing_extensions in Python 2,
+# so we can disallow importing it from anywhere except collections
+_BAD_COLLECTIONS_ALIASES = {
```
These names are kind of unclear, could you name them similarly to _TYPING_NOT_TYPING_EXTENSIONS with both the module they should be imported from and the module they should *not* be imported from?

AlexWaygood replied:
sure

AlexWaygood replied:
Done.

JelleZijlstra replied:
Thanks!

JelleZijlstra commented on pyi.py:
```diff
@@ -187,6 +239,76 @@ def in_class(self) -> bool:
+        thing = node.value
+        if not isinstance(thing, ast.Name):
+            return
+        thingname, attribute = thing.id, node.attr
```
```suggestion
        module_name, attribute = thing.id, node.attr
```
More precise variable name

AlexWaygood replied:
I went for `thingname` because we *don't* know that it's a module here -- this method visits all `ast.Attribute` nodes. It's only later down the method that we narrow it down as to whether it's an `ast.Attribute` we're interested in or not.

JelleZijlstra commented on pyi.py:
```diff
@@ -187,6 +239,76 @@ def in_class(self) -> bool:
+        elif object_name == "ContextManager":
+            error_code = Y023
+            blacklist = {
+                "ContextManager": '"contextlib.AbstractContextManager" or "typing.ContextManager"'
```
Shouldn't we pick one to recommend?

AlexWaygood replied:
`contextlib.AbstractContextManager` will fail on Python 2, typeshed still supports Python 2

JelleZijlstra replied:
I see, then maybe we should say something like `'"contextlib.AbstractContextManager" (or "typing.ContextManager" in Python 2-compatible code)'`. When a user sees this error message, I want them to have enough information to know how to fix it, and if we give them two names, how are they going to know which one to use?

AlexWaygood replied:
That makes sense... What do you think about the `collections.abc` case, though? `use "collections.abc.Awaitable", unless you stumble across a random mypy/pytype bug, in which case use `typing.Awaitable`?

JelleZijlstra replied:
I guess we should fix those bugs so this isn't an issue any more :) I don't recall the details of those bugs, are they common? Maybe we should just recommend typing until it's always safe to use collections.abc.

AlexWaygood replied:
> I guess we should fix those bugs so this isn't an issue any more :)

Indeed 😀

>  I don't recall the details of those bugs, are they common?

- https://github.com/python/mypy/issues/11860
- https://github.com/google/pytype/issues/1096

AlexWaygood replied:
TBF, the mypy issue looks typeshed-specific, and the pytype may be as well. But we certainly can't yet run a codemodding script using `collections.abc` everywhere in typeshed (if we could, I would have already done so ;)

JelleZijlstra commented on pyi.py:
```diff
@@ -187,6 +239,76 @@ def in_class(self) -> bool:
+    def _check_typing_object(
+        self, node: ast.Attribute | ast.ImportFrom, module_name: str, object_name: str
+    ) -> None:
+        if object_name in _BAD_COLLECTIONS_ALIASES:
```
This logic is a bit hard to follow because we don't always check the module name. You have to look at the other methods to see that this gets called only for typing/typing_extensions.

What if instead we always called _check_typing_object for each attribute and import, and then here had logic like
```
if module_name == "typing":
    if object_name not in _BAD_BUILTINS_ALIASES:
        return
    replacement = ...
elif module_name == "typing_extensions":
     # different logic for deciding if the name is bad
self.error(...)
```
That makes it easier to see at a glance which names will produce errors for each module.

AlexWaygood replied:
Okay!

JelleZijlstra commented on README.rst:
```diff
@@ -81,6 +81,16 @@ currently emitted:
+  ``typing_extensions``.
+* Y024: Use ``typing.NamedTuple`` instead of ``collections.namedtuple``, as it allows
+  for more precise type inference.
+* Y023: Always alias ``collections.abc.Set`` when importing it, so as to avoid
```
```suggestion
* Y025: Always alias ``collections.abc.Set`` when importing it, so as to avoid
```

AlexWaygood replied:
Good catch!

JelleZijlstra commented on pyi.py:
```diff
@@ -753,6 +881,14 @@ def should_warn(self, code):
+    'Y025 Use "from collections.abc import Set as AbstractSet" '
+    'to avoid confusion with "builtins.set"'
+)
+Y092 = "Y092 Top-level attribute must not have a default value"
```
```suggestion
```
Bad merge

JelleZijlstra commented on pyi.py:
```diff
@@ -187,6 +239,81 @@ def in_class(self) -> bool:
 
+    def _check_import_or_attribute(
+        self, node: ast.Attribute | ast.ImportFrom, module_name: str, object_name: str
+    ) -> None:
```
Thanks, this is much easier to read now

AlexWaygood commented:
I will get codemodding tomorrow!

AlexWaygood commented on pyi.py:
```diff
@@ -42,6 +42,58 @@ class TypeVarInfo(NamedTuple):
+    "DefaultDict": "defaultdict",
+    "ChainMap": "ChainMap",
+}
+_COLLECTIONS_NOT_TYPING_OR_TYPING_EXTENSIONS = {
```
Okay, I'll see if I can refactor the dicts. They used to have shorter variable names, but @JelleZijlstra persuaded me to make them longer ;)

I _hate_ the fact that the dict values include `"` characters, but it's the best way I could see of doing it. The complicating factor is the `contextlib.ContextManager` error message -- if you try to import it from `typing_extensions`, the inserted snippet is `"contextlib.AbstractContextManager" (or "typing.ContextManager" in Python 2-compatible code)`. So the inserted snippets are not uniformly surrounded by `"` characters.

AlexWaygood commented:
Typeshed PR to make the CI go green:

- https://github.com/python/typeshed/pull/6949

TODO:
- Try to refactor the dicts, per @Akuli's request
- Disallow importing `Type` from `typing_extensions` as well as `typing`

AlexWaygood commented:
Ohhhh I hardcoded Y019 to look for `type[_T]` instead of `Type[_T]`, so changing all occurences of `Type` to `type` in Python 2 has caused the CI to pick up on a bunch of new Y019 errors.

AlexWaygood commented:
flake8 is complaining that `PyiVisitor._check_import_or_attribute` is too complex for one function; however, I think it would make the code _more_ complex to split it up into several smaller functions, so I am going to `# noqa` it. Let me know if you disagree.

AlexWaygood commented:
> flake8 is complaining that `PyiVisitor._check_import_or_attribute` is too complex for one function. However, I think it would make the code _more_ complex to split it up into several smaller functions, so I am going to `# noqa` it. Let me know if you disagree.

Actually, I'll leave it for now, since I might be refactoring that function anyway.

AlexWaygood commented:
...And a third typeshed PR: https://github.com/python/typeshed/pull/6955

AlexWaygood commented:
CI is finally green, and I think I've addressed the points @Akuli raised (just two dicts now, and managed to get rid of the icky quotes-within-quotes everywhere). Re-requesting review from you, @JelleZijlstra, if that's okay, as it's changed a fair bit since when you approved it :)

Akuli reviewed (approved):
Skimmed. No red flags.

AlexWaygood commented:
> Skimmed. No red flags.

Thanks!
